src/uri/uri.cc

```diff
@@ -89,6 +92,54 @@ URI::URI(URI &&other)
   other.internal = nullptr;
 }
 
+auto URI::parse() -> void {
+  if (this->parsed) {
+    // clean
+    this->path_components_.clear();
+    this->scheme_ = std::nullopt;
+    this->userinfo_ = std::nullopt;
+    this->host_ = std::nullopt;
+    this->port_ = std::nullopt;
+    this->fragment_ = std::nullopt;
+    this->query_ = std::nullopt;
+    this->parsed = false;
+    uriFreeUriMembersA(&this->internal->uri);
+  }
+
+  uri_parse(this->data, &this->internal->uri);
+
+  const UriPathSegmentA *segment{this->internal->uri.pathHead};
+  if (segment != nullptr) {
+    // URNs and tags have a single path segment by definition
+    if (this->is_urn() || this->is_tag()) {
+      const auto part{uri_text_range(&segment->text)};
+      assert(part.has_value());
+      this->path_components_.push_back(std::string{part.value()});
+    } else {
+      while (segment) {
+        const auto part{uri_text_range(&segment->text)};
+        assert(part.has_value());
+        this->path_components_.push_back(std::string{part.value()});
+        segment = segment->next;
+      }
+    }
+  }
```

Comment:
It seems that for some reasons, `pathHead` points to the host.

Like in this test case:

```cpp
TEST(URI_path, without_scheme) {
  const sourcemeta::jsontoolkit::URI uri{"example.com/foo"};
  EXPECT_SPAN_EQ(uri.path().value(), {"foo"});
}
```

Which fails:

```text
/Users/tonygo/oss/jsontoolkit/test/uri/uri_path_test.cc:11: Failure
Expected equality of these values:
  span[i]
    Which is: "example.com"
  *it
    Which is: "foo"
Span element 0 does not match expected value
```

Note: I just catch the bug, the test is not pushed on this PR.

cc @jviotti

Reply:
I'll try to reproduce it on a standalone bug, to see what I missed here. It could be a silly side effect or something like that...

Reply:
I was able to reproduce it: 

```cpp
#include <iostream>
#include <string>
#include <uriparser/Uri.h>

auto uri_text_range(const UriTextRangeA *const range)
    -> std::optional<std::string_view> {
  if (range->afterLast == nullptr) {
    return std::nullopt;
  }

  return std::string_view{range->first,
                          static_cast<std::string_view::size_type>(
                              range->afterLast - range->first)};
}

auto print_path(const std::string uri) {
  UriUriA uri_struct;
  const char *error_position;
  uriParseSingleUriA(&uri_struct, uri.c_str(), &error_position);

  auto seg = uri_struct.pathHead;
  while (seg != nullptr) {
    const auto text = uri_text_range(&seg->text);
    if (text.has_value()) {
      std::cout << text.value() << std::endl;
    }
    seg = seg->next;
  }

  uriFreeUriMembersA(&uri_struct);
}

int main() {
  const std::string input_without_scheme = "example.com/foo/";
  const std::string input_with_scheme = "https://example.com/foo/bar/baz";

  print_path(input_without_scheme);
  print_path(input_with_scheme);
}
```

```
example.com
foo

foo
bar
baz
```

In one case we have the host, not in the other. Maybe it is expected. 🤔

Reply:
My 2 cents is that if the parser does not have enough context (the scheme) he probably deduce that is a path.

Reply:
I guess that I could update `normalize`/`canonicalize` to handle this case properly. For instance if I don't have a scheme in the URL then a `/` should not prefix the first path element.

Reply:
Hmmm to be honest I'm not sure. Before making assumptions or tweaking `normalize`/`canonicalize`, can you research what other popular parsers (maybe in JS, etc) do? Maybe it is indeed the expectation and we would have learnt something new.

Reply:
Or maybe even opening an issue on `uriparser` asking as well. URIs are very obscure, so I wouldn't be surprised if this feels weird for us but its actually compliant behavior, etc

Reply:
Yeah, I'll :)

Reply:
Quite interesting: https://datatracker.ietf.org/doc/html/rfc3986#section-4.5

Reply:
Yeah, very interesting. I also checked on the JSON Schema side and I think the current behavior is intended. i.e. if you just get `example.com/foo`, then it should be considered a path of `{ example.com, foo }` and we should not try to be smart and detect that `example.com` is the host

Reply:
Yeah for sure. What I was thinking about was to handle "/" for this specific case (we only have exception for URN and tag so far)

Lets go!